File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/user/usecase/AddSkinToStoreHistoryUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.user.usecase;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.ports.AuthenticatedUserProvider;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.ValorantAuthRepository;
import fr.huiitre.tools.modules.riot.valorant.application.user.command.AddSkinToStoreHistoryCommand;
import fr.huiitre.tools.modules.riot.valorant.application.user.ports.ValorantStoreHistoryRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.Optional;

@Service
@Transactional
public class AddSkinToStoreHistoryUseCase implements SecuredUseCase {

    private final AuthenticatedUserProvider authenticatedUserProvider;
    private final ValorantAuthRepository valorantAuthRepository;
    private final ValorantStoreHistoryRepository storeHistoryRepository;

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.USER;
    }

    public AddSkinToStoreHistoryUseCase(
            AuthenticatedUserProvider authenticatedUserProvider,
            ValorantAuthRepository valorantAuthRepository,
            ValorantStoreHistoryRepository storeHistoryRepository) {
        this.authenticatedUserProvider = authenticatedUserProvider;
        this.valorantAuthRepository = valorantAuthRepository;
        this.storeHistoryRepository = storeHistoryRepository;
    }

    public void execute(AddSkinToStoreHistoryCommand command) {
        Long accountId = command.getAccountId();
        if (!valorantAuthRepository.existsByIdAndUserId(accountId, authenticatedUserProvider.getUserId())) {
            throw new IllegalArgumentException("VALORANT_ACCOUNT_NOT_FOUND");
        }
        LocalDate seenAt = command.getSeenAt() != null ? command.getSeenAt() : LocalDate.now();

        if (command.getSkinIds() == null) return;

        for (Long skinId : command.getSkinIds()) {
            if (!storeHistoryRepository.existsByAccountIdAndSkinIdAndDate(accountId, skinId, seenAt)) {
                storeHistoryRepository.add(accountId, skinId, seenAt);
            }
        }
    }
}

File: api/Modules/Riot/Valorant/Application/User/Usecases/AddSkinToStoreHistoryUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Common.Application.Ports;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.User.Commands;
using Tools.Api.Modules.Riot.Valorant.Application.User.Ports;

namespace Tools.Api.Modules.Riot.Valorant.Application.User.Usecases;

// Archive la rotation du jour d'un coup : le front envoie toute la boutique en un appel.
public sealed class AddSkinToStoreHistoryUseCase(
    UseCaseAuthorizer authorizer,
    IValorantAuthRepository valorantAuthRepository,
    IValorantStoreHistoryRepository storeHistoryRepository,
    ITransactionManager transactionManager
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.User;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    public async Task Execute(AddSkinToStoreHistoryCommand command)
    {
        if (!await valorantAuthRepository.ExistsByIdAndUserId(command.AccountId, CurrentUser.UserId))
        {
            throw AppException.NotFound(
                "VALORANT_ACCOUNT_NOT_FOUND",
                "Ce compte Valorant est introuvable.");
        }

        if (command.SkinIds is not { Count: > 0 } skinIds)
        {
            return;
        }

        // Le Java retombait sur la date du jour quand l'appelant n'en donnait pas.
        var seenAt = command.SeenAt ?? DateOnly.FromDateTime(DateTime.UtcNow);

        // Le lot est indivisible : une rotation à moitié archivée serait relue comme incomplète.
        await using var transaction = await transactionManager.BeginAsync();

        foreach (var skinId in skinIds)
        {
            if (!await storeHistoryRepository.ExistsByAccountIdAndSkinIdAndDate(
                    command.AccountId, skinId, seenAt))
            {
                await storeHistoryRepository.Add(command.AccountId, skinId, seenAt);
            }
        }

        await transaction.CommitAsync();
    }
}
